packages/map-toolkit/src/viewport/use-viewport-state.ts

```diff
@@ -0,0 +1,221 @@
+/*
+ * Copyright 2025 Hypergiant Galactic Systems Inc. All rights reserved.
+ * This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+import { Broadcast } from '@accelint/bus';
+import { type UniqueId, uuid } from '@accelint/core';
+import { useSyncExternalStore } from 'react';
+import { MapEvents } from '../deckgl/base-map/events';
+import type {
+  MapEventType,
+  MapViewportPayload,
+} from '../deckgl/base-map/types';
+
+const bus = Broadcast.getInstance<MapEventType>();
+
+export type UseViewportStateProps = {
+  instanceId: UniqueId;
+  subscribe?: Parameters<typeof useSyncExternalStore<MapViewportPayload>>[0];
+  getSnapshot?: Parameters<typeof useSyncExternalStore<MapViewportPayload>>[1];
+  getServerSnapshot?: Parameters<
+    typeof useSyncExternalStore<MapViewportPayload>
+  >[2];
+};
+
+/**
+ * Store for viewport state keyed by instanceId
+ */
+const viewportStore = new Map<UniqueId, MapViewportPayload>();
+
+/**
+ * Track the number of active subscribers per instanceId for automatic cleanup
+ */
+const subscriberCounts = new Map<UniqueId, number>();
```

Comment:
Per Andrew's catch, separate component subscribers from map instance subscriptions:


```suggestion
/**
 * Track React component subscribers per instanceId (for fan-out notifications).
 * Each Set contains onStoreChange callbacks from useSyncExternalStore.
 */
const componentSubscribers = new Map<UniqueId, Set<() => void>>();

/**
 * Cache of bus unsubscribe functions (1 per instanceId).
 * This ensures we only have one bus listener per viewport, regardless of
 * how many React components subscribe to it.
 */
const busUnsubscribers = new Map<UniqueId, () => void>();
```